### PR TITLE
Optimizing equal and compare when their arguments are referencially equal

### PR DESCRIPTION
**Pull request by Pauan**
The GitHub diff is bad. Look at the diff in something like [meld](https://en.wikipedia.org/wiki/Meld_(software)) instead (with whitespace ignored).

### CODE REVIEW

**Comment by jims** [comment]
You can also add `?w=1` to the url to ignore whitespace :)

**Comment by Pauan** [comment]
@jims That's super useful! That should be a checkbox on the "Files changed" page.

File: src/fable/Fable.Core/ts/List.ts [hunk 1]
```diff
@@ -121,30 +121,42 @@ export default class List<T> implements IEquatable<List<T>>, IComparable<List<T>
...
-        return cur2.done ? true : false;
-      else if (cur2.done)
-        return false;
-      else if (!equals(cur1.value, cur2.value))
-        return false
+    // Optimization if they are referencially equal
+    if (this === x) {
+      return true;
+
+    } else {
```
Comment (jims) [inline src/fable/Fable.Core/ts/List.ts]: Isn't this `else` superflous?

**Comment by tpetricek** [comment]
I was thinking the other day that having this optimization would be nice too! It seems to be breaking some tests though - can you look into that and figure out what's going on there?

**Comment by alfonsogarciacaro** [comment]
This has actually surfaced a [bug in Date comparison](https://github.com/fable-compiler/Fable/commit/2d8342859bdef6742aca3a605d98493cf5798e6f#diff-91d3a972bb20dc75c94fd935b1f58fb6R154), thank you! :+1:

**Comment by et1975** [comment]
Can this be back ported to 0.6.x?

**Comment by Pauan** [comment]
@et1975 It's possible, but the structure of `fable-core` has changed between `0.6` and `0.7`, so it would require a non-trivial amount of work: I don't think it's as simple as using `git cherry-pick`. Is there any particular reason you can't use `0.7`?

**Comment by et1975** [comment]
0.7.x entails a large amount of unknown, and I don't know when we'll be able to move, but fair enough. I thought it might be non-trivial (hey, look at that record equality difference between the two!) figured I'll ask before backporting it myself.

**Pull request merged**